File: examples/java/kora-java-database-jdbc/src/main/java/io/koraframework/example/jdbc/JdbcIdSequenceRepository.java
package io.koraframework.example.jdbc;

import org.jspecify.annotations.Nullable;
import java.util.List;
import io.koraframework.database.common.annotation.*;
import io.koraframework.database.jdbc.annotation.EntityJdbc;
import io.koraframework.database.jdbc.JdbcRepository;

@Repository
public interface JdbcIdSequenceRepository extends JdbcRepository {

    @EntityJdbc
    record Entity(@Id Long id,
                  @Column("name") String name) {

        public Entity(String name) {
            this(null, name);
        }
    }

    @Query("SELECT * FROM entities_sequence WHERE id = :id")
    @Nullable
    Entity findById(long id);

    @Query("""
            INSERT INTO entities_sequence(name)
            VALUES (:entity.name)
            RETURNING id
            """)
    long insert(Entity entity);

    @Query("""
            INSERT INTO entities_sequence(name)
            VALUES (:entity.name)
            """)
    @Id
    Long insertGenerated(Entity entity);

    @Query("""
            INSERT INTO entities_sequence(name)
            VALUES (:entity.name)
            """)
    @Id
    List<Long> insertGenerated(@Batch List<Entity> entity);
}

File: examples/kotlin/kora-kotlin-database-jdbc/src/main/kotlin/io/koraframework/kotlin/example/jdbc/JdbcIdSequenceRepository.kt
package io.koraframework.kotlin.example.jdbc

import org.postgresql.util.PGobject
import io.koraframework.common.annotation.Component
import io.koraframework.common.annotation.Mapping
import io.koraframework.common.annotation.Module
import io.koraframework.common.annotation.Root
import io.koraframework.database.common.UpdateCount
import io.koraframework.database.common.annotation.*
import io.koraframework.database.jdbc.annotation.EntityJdbc
import io.koraframework.database.jdbc.JdbcRepository
import io.koraframework.database.jdbc.mapper.parameter.JdbcParameterColumnMapper
import io.koraframework.database.jdbc.mapper.result.JdbcResultColumnMapper
import io.koraframework.database.jdbc.mapper.result.JdbcResultSetMapper
import io.koraframework.database.jdbc.mapper.result.JdbcRowMapper
import io.koraframework.json.common.JsonReader
import io.koraframework.json.common.JsonWriter
import io.koraframework.json.common.annotation.Json
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.sql.Types

@Repository
interface JdbcIdSequenceRepository : JdbcRepository {

    @EntityJdbc
    data class Entity(@field:Id val id: Long?, @field:Column("name") val name: String) {
        constructor(name: String) : this(null, name)
    }

    @Query("SELECT * FROM entities_sequence WHERE id = :id")
    fun findById(id: Long): Entity?

    @Query("INSERT INTO entities_sequence(name) VALUES (:entity.name) RETURNING id")
    fun insert(entity: Entity): Long

    @Id
    @Query("INSERT INTO entities_sequence(name) VALUES (:entity.name)")
    fun insertGenerated(entity: Entity): Long

    @Id
    @Query("INSERT INTO entities_sequence(name) VALUES (:entity.name)")
    fun insertGenerated(@Batch entity: List<Entity>): List<Long>
}


